Automattic/woocommerce-payments
# Fix losing cart contents during the login at checkout

Fixes #7992

### Changes proposed in this Pull Request

Narrow down to specific conditions when the `WC_Payment_Gateway_WCPay::set_cookie_on_current_request` filter is applied. 

- During a checkout process request. `WOOCOMMERCE_CHECKOUT` is an indicator for this request, which is supported in both classic and block checkouts. 
- A new account is created during this process. Ensure that action `woocommerce_created_customer` is called out at least one time. This is also supported in both classic and block checkouts. This check also supports multi cases of creating a new account during the checkout process request: optional or automatic account creation (for a checkout including a subscription product). 

Why not using the condition of 3DS? 

- This can be only possible to know after sending the request of creating payment intent. It's [somewhere here in the code](https://github.com/Automattic/woocommerce-payments/blob/c9078890fc0f2394d6c097c90bad2fb5717fe3d1/includes/class-wc-payment-gateway-wcpay.php#L1741-L1764). We can inject a filter around this code, but it will complicate the code a lot.
- Another important factor is that the process of creating a new account is run before the payment process. Therefore, until we know a payment requiring 3DS, it's too late. 









#### Testing instructions

- Ensure to add the snippet mentioned in this issue: https://github.com/Automattic/woocommerce-payments/issues/7992 . It's possible to use  plugin Code Snippet for this purpose, otherwise, just add this snippet to the bottom of the main file `woocommerce-payments.php`.
- Test 1: follow `To Reproduce` https://github.com/Automattic/woocommerce-payments/issues/7992
- Expectation test 1: the cart is persisted. 
- Test 2: follow manual instruction testings in https://github.com/Automattic/woocommerce-payments/pull/819
- Expectation test 2: 3DS payment is run through without any issue. 

Ideally, these flows would be tested for both types of checkout: block and classic.





*

-------------------


**Post merge**
